src/mapnode.cpp

```diff
@@ -170,64 +170,158 @@ void MapNode::rotateAlongYAxis(const NodeDefManager *nodemgr, Rotation rot)
-				box.MaxEdge.rotateYZBy(-90);
-				break;
-			case 3:  //x+
+			break;
+		case 1: // z+
+			box.MinEdge.rotateYZBy(90);
+			box.MaxEdge.rotateYZBy(90);
+			if(facedir == 1)
+			{
+				box.MinEdge.rotateXYBy(90);
+				box.MaxEdge.rotateXYBy(90);
+			}
+			else if(facedir == 2)
+			{
+				box.MinEdge.rotateXYBy(180);
+				box.MaxEdge.rotateXYBy(180);
+			}
+			else if(facedir == 3)
+			{
 				box.MinEdge.rotateXYBy(-90);
 				box.MaxEdge.rotateXYBy(-90);
-				break;
-			case 4:  //x-
+			}
+			break;
+		case 2: //z-
+			box.MinEdge.rotateYZBy(-90);
+			box.MaxEdge.rotateYZBy(-90);
+			if(facedir == 1)
+			{
+				box.MinEdge.rotateXYBy(-90);
+				box.MaxEdge.rotateXYBy(-90);
+			}
+			else if(facedir == 2)
+			{
+				box.MinEdge.rotateXYBy(180);
+				box.MaxEdge.rotateXYBy(180);
+			}
+			else if(facedir == 3)
+			{
 				box.MinEdge.rotateXYBy(90);
 				box.MaxEdge.rotateXYBy(90);
-				break;
-			case 5:
-				box.MinEdge.rotateXYBy(-180);
-				box.MaxEdge.rotateXYBy(-180);
-				break;
-			default:
-				break;
 			}
+			break;
+		case 3:  //x+
+			box.MinEdge.rotateXYBy(-90);
+			box.MaxEdge.rotateXYBy(-90);
+			if(facedir == 1)
+			{
+				box.MinEdge.rotateYZBy(90);
+				box.MaxEdge.rotateYZBy(90);
+			}
+			else if(facedir == 2)
+			{
+				box.MinEdge.rotateYZBy(180);
+				box.MaxEdge.rotateYZBy(180);
+			}
+			else if(facedir == 3)
+			{
+				box.MinEdge.rotateYZBy(-90);
+				box.MaxEdge.rotateYZBy(-90);
+			}
+			break;
+		case 4:  //x-
+			box.MinEdge.rotateXYBy(90);
+			box.MaxEdge.rotateXYBy(90);
+			if(facedir == 1)
+			{
+				box.MinEdge.rotateYZBy(-90);
+				box.MaxEdge.rotateYZBy(-90);
+			}
+			else if(facedir == 2)
+			{
+				box.MinEdge.rotateYZBy(180);
+				box.MaxEdge.rotateYZBy(180);
+			}
+			else if(facedir == 3)
+			{
+				box.MinEdge.rotateYZBy(90);
+				box.MaxEdge.rotateYZBy(90);
+			}
+			break;
+		case 5:
+			box.MinEdge.rotateXYBy(-180);
+			box.MaxEdge.rotateXYBy(-180);
+			if(facedir == 1)
+			{
+				box.MinEdge.rotateXZBy(90);
+				box.MaxEdge.rotateXZBy(90);
+			}
+			else if(facedir == 2)
+			{
+				box.MinEdge.rotateXZBy(180);
+				box.MaxEdge.rotateXZBy(180);
+			}
+			else if(facedir == 3)
+			{
+				box.MinEdge.rotateXZBy(-90);
+				box.MaxEdge.rotateXZBy(-90);
+			}
+			break;
+		default:
+			break;
+		}
+		box.repair();
+		boxes.push_back(box);
+	}
+}
 
-			box.repair();
-			boxes.push_back(box);
+void transformNodeBox(const MapNode &n, const NodeBox &nodebox,
+	const NodeDefManager *nodemgr, std::vector<aabb3f> *p_boxes,
+	u8 neighbors = 0)
+{
+	std::vector<aabb3f> &boxes = *p_boxes;
+
+	if (nodebox.type == NODEBOX_FIXED || nodebox.type == NODEBOX_LEVELED) {
+		const std::vector<aabb3f> &fixed = nodebox.fixed;
+		const std::vector<aabb3f> &leveled_fixed = nodebox.leveled_fixed;
+		enum NodeBoxType nbt = nodebox.type;
+		if (nbt == NODEBOX_LEVELED){
+			buildFixedNodeBox(n, nodebox, nodemgr, boxes, leveled_fixed, NODEBOX_FIXED);
 		}
+		buildFixedNodeBox(n, nodebox, nodemgr, boxes, fixed, nbt);
```

Comment:
```suggestion
		if (nodebox.type == NODEBOX_LEVELED) {
			// add leveled_fixed boxes as if they were really fixed (because we modify the fixed ones...)
			buildFixedNodeBox(n, nodebox, nodemgr, boxes, nodebox.leveled_fixed, NODEBOX_FIXED);
		}
		buildFixedNodeBox(n, nodebox, nodemgr, boxes, nodebox.fixed, nodebox.type);
```